File: 0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */

function maxSubArray(nums) {
    const n = nums.length
    let maxSum = -Infinity
    let sum = 0
    
    for (let i = 0; i < n; i++) {
        sum = Math.max(nums[i], sum + nums[i])
        maxSum = Math.max(sum, maxSum)
    }
    return maxSum
}

// -2 1 -3

// sum 

File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = nums[0]
        cur = 0
        
        for n in nums:
            if cur < 0:
                cur = 0
            cur += n
            maxSum = max(maxSum, cur)
        return maxSum